File: app/src/renderer/types/query.ts
export interface QueryRequest {
  connection_id: string
  sql: string
  bind_variables?: Record<string, unknown>
  limit?: number
  offset?: number
  timeout_seconds?: number
}

export interface ColumnMeta {
  name: string
  type: string
  nullable: boolean
}

export interface QueryResult {
  execution_id: string
  columns: ColumnMeta[]
  rows: unknown[][]
  row_count: number
  total_count: number | null
  duration_ms: number
  truncated: boolean
}

export interface ExplainResult {
  raw_plan: unknown
  summary: {
    total_cost: number | null
    execution_time_ms: number | null
    nodes: unknown[]
  }
}

export interface HistoryEntry {
  id: string
  connection_id: string
  sql_text: string
  dialect: string
  status: string
  error_message: string | null
  row_count: number | null
  duration_ms: number | null
  executed_at: string
}

File: crates/ferrite-db/src/types/query.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnMeta {
    pub name: String,
    #[serde(rename = "type")]
    pub data_type: String,
    pub nullable: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryResult {
    pub execution_id: Uuid,
    pub columns: Vec<ColumnMeta>,
    pub rows: Vec<Vec<serde_json::Value>>,
    pub row_count: usize,
    pub total_count: Option<usize>,
    pub duration_ms: u64,
    pub truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExplainResult {
    pub raw_plan: serde_json::Value,
    pub summary: ExplainSummary,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExplainSummary {
    pub total_cost: Option<f64>,
    pub execution_time_ms: Option<f64>,
    pub nodes: Vec<ExplainNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExplainNode {
    pub node_type: String,
    pub relation: Option<String>,
    pub cost: Option<f64>,
    pub rows: Option<u64>,
    pub width: Option<u64>,
    pub children: Vec<ExplainNode>,
}
